### PR TITLE
Fix tasks becoming unordered, fixes #7645

### LINKED ISSUE #7645: tasks become unsorted - task list sort order is not preserved (tasksOrder errors)
**Sometimes when you sort your To-Dos, some or all of them will become unsorted, either immediately after you have sorted them, or later when you add new To-Dos.**

**The workaround is to post to the [Report a Bug guild](https://habitica.com/#/options/groups/guilds/a29da26b-37de-4a71-b0c6-48e72a900dac) to say that your To-Do sort order is not being saved. An admin will then cleanse some data in your account which will fix the problem. You cannot fix this problem yourself. In particular, deleting completed To-Dos will NOT help (it used to help for a previous version of this bug, but it won't make any difference now).**

It is likely that this bug occurs only during periods of heavy site use, or if you mark several To-Dos as completed very quickly, or if you mark one as completed at the same time as taking other actions on the site very quickly. To minimise the chance of this bug happening (and certain other bugs and sync errors):
- on the website, each time you take an action, wait for the blue sync line at the top of the site to disappear before you take another action
- on the mobile app, each time you mark a task as complete, don't take any other actions until you've seen the notification at the bottom of the screen that tells you how much XP and GP you earned
### Technical Details

---

_EDIT by @paglias_: the best place to fix this is at cron. It can also happen that some tasks ids are missing from user.tasksOrder, this should be fixed too.

---

When you mark a To-Do as completed, Habitica has to modify the To-Do's record and also remove the To-Do's ID from a list that stores the sort order of your incomplete To-Dos (`tasksOrder.todos`). If Habitica is overloaded at the time, it might not do that second action. Thus the sort-order list can contain IDs for completed To-Dos, which interfere when you try to sort your incomplete To-Dos.

This was found when PaperCrystals (52e31d82-e731-4713-81cb-6ef8c5d5d9bc) reported problems with sorting To-Dos in the Report a Bug guild. PaperCrystals's `tasksOrder.todos` array contained the IDs of five To-Dos that had been completed. Their completion dates were all within 64 seconds, indicating that there might have been problems with the user document being updated multiple times in a short period:

```
"dateCompleted" : ISODate("2016-05-28T19:48:28.021Z")
"dateCompleted" : ISODate("2016-05-28T19:48:46.384Z")
"dateCompleted" : ISODate("2016-05-28T19:49:00.962Z")
"dateCompleted" : ISODate("2016-05-28T19:49:23.729Z")
"dateCompleted" : ISODate("2016-05-28T19:49:32.054Z")
```

I was able to replicate the bug by inserting fake task IDs into my own `tasksOrder.todos` array. To-Do sorting then became impossible: I could move a To-Do to a new position, but as soon as I clicked sync, it would jump to a different position.

To find out how widespread this problem might be, I am searching for completed To-Dos in the `tasksOrder.todos` array for all players who logged in in the past three months (since 2016-03-15). It's going to take a while for that to finish, but initial results indicate that maybe 2% of the IDs in the `tasksOrder.todos` arrays are for completed To-Dos.

-------

With PR #12335 the Habitica codebase has been updated to support MongoDB 4.2 and transactions making it possible to fix this issue. 

Transactions allow operations on multiple documents to be executed ensuring that either all of them are executed correctly or none, making it possible to fix this issue.

An example on using transactions can be found at https://mongoosejs.com/docs/transactions.html while more info at https://docs.mongodb.com/manual/core/transactions/. If you want to work on this issue and have any question please leave a comment!

### PR DESCRIPTION
**Pull request by TNychka**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Partial fix for https://github.com/HabitRPG/habitica/issues/7645

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
+ On cron, filter out any completed todos from the taskOrder list
+ Ignore any completed todos when getting the order of tasks from server

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: f57bd235-ec21-4357-b5c9-a0536294b285

### CODE REVIEW

File: website/server/libs/cron.js [hunk 1]
```diff
@@ -403,6 +403,11 @@ export function cron (options = {}) {
 
   user.history.exp.push({date: now, value: expTally});
 
+  // Remove any remaining completed todos from the list of active todos
+  user.tasksOrder.todos = tasksByType.todos.filter(task => {
```
Comment (paglias) [inline website/server/libs/cron.js]: this should return ids not the entire tasks

Comment (TNychka) [reply website/server/libs/cron.js]: Oh right, that's a silly mistake. I'll fix that

File: website/server/libs/taskManager.js [hunk 2]
```diff
@@ -199,10 +199,19 @@ export async function getTasks (req, res, options = {}) {
   // Order tasks based on tasksOrder
   if (type && type !== 'completedTodos' && type !== '_allCompletedTodos') {
     let order = owner.tasksOrder[type];
-    let orderedTasks = new Array(tasks.length);
+
+    // Remove possible completed todos in the task list
```
Comment (paglias) [inline website/server/libs/taskManager.js]: Given the query here https://github.com/HabitRPG/habitica/pull/8928/files#diff-cb32c22fd9c0bad615b785ed3f988796R182 that excludes completed todos, I don't think it's possible for completed todos to be inside the array

**Review by paglias — DISMISSED** [review]

File: website/server/libs/cron.js (hunk 1 above)
Comment (TNychka) [reply website/server/libs/cron.js]: Fixed

File: website/server/libs/taskManager.js (hunk 2 above)
Comment (TNychka) [reply website/server/libs/taskManager.js]: Removed this, I inspected this and you're right it it probably irrelevant

**Comment by TNychka** [comment]
@paglias I think this is ready for another review. I added some tests and re-wrote what was there before.

**Comment by paglias** [comment]
@TNychka thanks for the update but I think the same problem as before remains: `tasksByType` is created from the database making sure completed todos aren't included, see https://github.com/HabitRPG/habitica/blob/develop/website/server/middlewares/cron.js#L60 so it's not possible to have completed todos in tasksByType in `libs/cron`

**Comment by TNychka** [comment]
Right, but the issue is that user.tasksOrder.todos can sometimes contain completed todos. So I use the tasksByType list which does not contain any completed todos to filter the user.tasksOrder.todos which could contain the ids of some completed todos.

**Comment by paglias** [comment]
@TNychka that's a good idea but you're not doing that. See this if https://github.com/HabitRPG/habitica/pull/8928/files#diff-95ea51d6c9aaabaf272f7ff354e1d8a5R408, if tasksByType never contains a completed task it'll never evalua

**Comment by TNychka** [comment]
Oh man, I see it now. I'll fix that.. Again. Thanks I'll let you know

**Comment by TNychka** [comment]
@paglias Okay, I hope this is alright. Correct my logic if I'm wrong.
tasksByType does NOT contain any completed todos
user.tasksOrder.todos MAY contain some ids of completed todos
In order to fix this, we can just ensure that any task ids in the tasksOrder are also in the tasksByType.todos.
Order of todos should be preserved.
If this is the case then the above code does fix the issue. I did leave in the check to see if the task was complete since it really does not add any extra complexity.

**Comment by paglias** [comment]
@TNychka that looks like the correct approach! Unfortunately we're in the middle of a big front end redesign so for the time being we're not reviewing or merging PRs, we will start again as soon as the new client is launched which hopefully means in less than 2 week. I hope you understand

**Comment by TNychka** [comment]
Hey, just checking to see if that merge is complete!

**Comment by Alys** [comment]
@TNychka When a pull request has been merged, the green "Open" button at the top changes to a purple "Merged" button, so this hasn't been merged yet. You can see an example at https://github.com/HabitRPG/habitica/pull/9067 We'll be reviewing PRs again after the website's front-end redesign is finished.

**Comment by TNychka** [comment]
Hey, just checking if PRs are good to go again or if you're still working on the front end changes

**Comment by lemoness** [comment]
The frontend changes have been completed, so we are now accepting PRs again! We're prioritizing bug fixes, but will be reviewing older PRs starting this week.

**Comment by TNychka** [comment]
Rebased to dev head

**Comment by SabreCat** [comment]
Hey @TNychka, that rebase made it very difficult to tell what was actually changed on this PR. (Merges usually have better results, when possible!) What files are relevant? `website/server/libs/cron.js`, `website/server/libs/taskManager.js`, anything else?

**Comment by SabreCat** [comment]
Yeah, I don't think we can merge this PR with that many diffs. Can you create a fresh branch off `develop`, reapply your changes, and submit a new PR? Alternatively, let me know the list of files whose changes are relevant to the PR, and I'll see if I can reconstruct it myself.

**Comment by TNychka** [comment]
I have no idea what happened to create all those diffs. Fixed now, @SabreCat PTAL

**Review by SabreCat — APPROVED** [review]

**Comment by SabreCat** [comment]
Going to let tests finish for the fixed-up PR before merging, but this looks good to me!

**Comment by SabreCat** [comment]
Thanks much for this, @TNychka! I've added this to your contributor credits toward Tier 2 Blacksmith!

**Pull request merged**